chromium_src/chrome/browser/ui/search_engines/search_engine_tab_helper.cc

```diff
@@ -0,0 +1,16 @@
+/* Copyright 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "chrome/browser/profiles/profile_manager.h"
+#include "components/prefs/pref_service.h"
+#include "components/search_engines/search_engines_pref_names.h"
+
+#define PREVENT_OSDD_ADDITION                                                \
+  if (!profile->GetPrefs()->GetBoolean(prefs::kAddOpenSearchEngines)) { \
```

Comment:
Seems like this check can be done at the very start of both functions thus we can avoid creating ```.patch``` for that.
It's okay to get ```Profile*``` twice, it won't hurt performance.
Example of how it can be done can be found here: https://github.com/brave/brave-core/pull/9432/files ```cached_storage_area.h/cached_storage_area.cc```

Reply:
I tried that approach, but it didn't work. `SearchEngineTabHelper` instantiates the `WebContentsUserData` template (see [here](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/search_engines/search_engine_tab_helper.h;drc=ffe737cf5019ea200028b2917983e00003ff3425;l=27)) in it's constructor, which is how an instance of this class is retrieved (see [here](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/web_contents_user_data.h;drc=d71b784f527b3a422850a53787662a9722d465b1;l=64)).

Because of this, the derived class never gets constructed. If I inherit from `WebContentsUserData` again in the derived class, that would create two instances of `WebContentsUserData` and introduce ambiguity. I'm not sure if it's possible to use virtual overrides without creating a `.patch` again.

Reply:
So I tried that approach but it didn't work. The `WebContentsUserData` template is instantiated when `SearchEngineTabHelper` is constructed (see [here](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/search_engines/search_engine_tab_helper.h;drc=ffe737cf5019ea200028b2917983e00003ff3425;l=27)), and that is how an instance of this class is retrieved (see [here](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/web_contents_user_data.h;drc=d71b784f527b3a422850a53787662a9722d465b1;l=47)). 

Because of this the derived class is never constructed. If we inherit `WebContentsUserData` in the derived class again, that would create two instances and introduce ambiguity. I'm not sure how we can use virtual overrides without creating a `.patch` again.

Reply:
Gotcha. What if we create ```SearchEngineTabHelper::GenerateKeywordFromNavigationEntry``` and check for this pref inside and return an empty string if the pref is not set? If the pref is fine then just call ```::GenerateKeywordFromNavigationEntry(...)``` next. Something like that:
```
std::u16string SearchEngineTabHelper::GenerateKeywordFromNavigationEntry(
    NavigationEntry* entry) {
  Profile* profile =
      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
  if (!profile->GetPrefs()->GetBoolean(prefs::kAddOpenSearchEngines)) {
    return std::u16string();
  }
  return ::GenerateKeywordFromNavigationEntry(entry);
}
```

Reply:
Yep that works, thanks :) 

Updated.

Reply:
@goodov I just noticed that upstream is introducing `GenerateKeywordFromNavigationEntry` in `SearchEngineTabHelper` starting from [93.0.4548.0](https://source.chromium.org/chromium/chromium/src/+/refs/tags/93.0.4548.0:chrome/browser/ui/search_engines/search_engine_tab_helper.cc;bpv=0;bpt=0). This code will break in that case.

Reply:
what about doing the same trick in ```SearchEngineTabHelper::IsFormSubmit``` ?

Reply:
That works! Updated.